src/packet/sym_encrypted_integrity_protected_data.js

```diff
@@ -113,33 +150,148 @@ class SymEncryptedIntegrityProtectedDataPacket {
-    const decrypted = await crypto.mode.cfb.decrypt(sessionKeyAlgorithm, key, encrypted, new Uint8Array(blockSize));
-
-    // there must be a modification detection code packet as the
-    // last packet and everything gets hashed except the hash itself
-    const realHash = stream.slice(stream.passiveClone(decrypted), -20);
-    const tohash = stream.slice(decrypted, 0, -20);
-    const verifyHash = Promise.all([
-      stream.readToEnd(await crypto.hash.sha1(stream.passiveClone(tohash))),
-      stream.readToEnd(realHash)
-    ]).then(([hash, mdc]) => {
-      if (!util.equalsUint8Array(hash, mdc)) {
-        throw new Error('Modification detected.');
+
+    let packetbytes;
+    if (this.version === 2) {
+      packetbytes = await this.aead('decrypt', key, encrypted);
+    } else {
+      const { blockSize } = crypto.getCipher(sessionKeyAlgorithm);
+      const decrypted = await crypto.mode.cfb.decrypt(sessionKeyAlgorithm, key, encrypted, new Uint8Array(blockSize));
+
+      // there must be a modification detection code packet as the
+      // last packet and everything gets hashed except the hash itself
+      const realHash = stream.slice(stream.passiveClone(decrypted), -20);
+      const tohash = stream.slice(decrypted, 0, -20);
+      const verifyHash = Promise.all([
+        stream.readToEnd(await crypto.hash.sha1(stream.passiveClone(tohash))),
+        stream.readToEnd(realHash)
+      ]).then(([hash, mdc]) => {
+        if (!util.equalsUint8Array(hash, mdc)) {
+          throw new Error('Modification detected.');
+        }
+        return new Uint8Array();
+      });
+      const bytes = stream.slice(tohash, blockSize + 2); // Remove random prefix
+      packetbytes = stream.slice(bytes, 0, -2); // Remove MDC packet
+      packetbytes = stream.concat([packetbytes, stream.fromAsync(() => verifyHash)]);
+      if (!util.isStream(encrypted) || !config.allowUnauthenticatedStream) {
+        packetbytes = await stream.readToEnd(packetbytes);
       }
-      return new Uint8Array();
-    });
-    const bytes = stream.slice(tohash, blockSize + 2); // Remove random prefix
-    let packetbytes = stream.slice(bytes, 0, -2); // Remove MDC packet
-    packetbytes = stream.concat([packetbytes, stream.fromAsync(() => verifyHash)]);
-    if (!util.isStream(encrypted) || !config.allowUnauthenticatedStream) {
-      packetbytes = await stream.readToEnd(packetbytes);
     }
+
     this.packets = await PacketList.fromBinary(packetbytes, allowedPackets, config);
     return true;
   }
+
+  /**
+   * En/decrypt the payload.
+   * NOTE: this function is also used by the AEADEncryptedDataPacket.
+   * @param {encrypt|decrypt} fn - Whether to encrypt or decrypt
+   * @param {Uint8Array} key - The session key used to en/decrypt the payload
+   * @param {Uint8Array | ReadableStream<Uint8Array>} data - The data to en/decrypt
+   * @returns {Promise<Uint8Array | ReadableStream<Uint8Array>>}
+   * @async
+   */
+  async aead(fn, key, data) {
```

Comment:
Can this be a non-class function that simply takes a `packet: AEADEncryptedDataPacket | SymEncryptedIntegrityProtectedDataPacket` argument, and gets called by both classes? So we don't need to use `.apply` to reuse it. Also, doesn't `instanceof` work?

Reply:
It's possible, but I don't want to change the code to accommodate the `AEADEncryptedDataPacket` use case too much; I would rather that the `SymEncryptedIntegrityProtectedDataPacket` is most easy to read, and `AEADEncryptedDataPacket` piggy-backs off of that. That way, if we can ever drop support for `AEADEncryptedDataPacket`, the code can stay essentially the same.
And, `instanceof` also works but it walks the entire prototype chain, which in this case I think isn't necessary since we don't have subclasses of `SymEncryptedIntegrityProtectedDataPacket` and `AEADEncryptedDataPacket`, so we can just check the constructor directly; that might also be slightly faster.